Add network interface settings for mDNS/LLMNR

t3hk0d3 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change

Currently there is an issue with mDNS announcement enabled by default on all interfaces.
See https://github.com/home-assistant/supervisor/issues/5435 and https://github.com/home-assistant/operating-system/issues/1163

This PR is a second take to fix this issue.
First attempt https://github.com/home-assistant/supervisor/pull/5455 was using heuristic logic to enable mDNS annoucement only for primary interface.
After some discusison with @agners we decided that's not an unacceptable solution, because it would break mDNS for people without installed mDNS reflector on router.

This PR take a different route to solve this issue:
- reads status of mDNS/LLMNR from D-Bus
- if status is not set: it will fallback to current behaviour - set it to `announce`.
- exposes these new properties to API

<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

## Type of change

<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (which adds functionality to the supervisor)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information

<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes https://github.com/home-assistant/supervisor/issues/5435 and https://github.com/home-assistant/operating-system/issues/1163
- This PR is related to issue:
- Link to documentation pull request:
- Link to cli pull request: https://github.com/home-assistant/cli/pull/591
- Link to client library pull request: https://github.com/home-assistant-libs/python-supervisor-client/pull/173

## Checklist

<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Ruff (`ruff format supervisor tests`)
- [x] Tests have been added to verify that the new code works.

If API endpoints or add-on configuration are added/changed:

- [ ] Documentation added/updated for [developers.home-assistant.io][docs-repository]
- [x] [CLI][cli-repository] updated (if necessary)
- [x] [Client library][client-library-repository] updated (if necessary)

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[docs-repository]: https://github.com/home-assistant/developers.home-assistant
[cli-repository]: https://github.com/home-assistant/cli
[client-library-repository]: https://github.com/home-assistant-libs/python-supervisor-client/


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Release Notes

- **New Features**
  - Added support for configuring Multicast DNS (mDNS) and Link-Local Multicast Name Resolution (LLMNR) settings for network interfaces.
  - Introduced new configuration options to control mDNS and LLMNR behavior (OFF, RESOLVE, ANNOUNCE).

- **Improvements**
  - Enhanced network interface configuration API to include mDNS and LLMNR settings.
  - Expanded network connection properties to support more detailed network service discovery options.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

sairon commented on supervisor/host/configuration.py:
```diff
@@ -145,6 +154,13 @@ def from_dbus_interface(inet: NetworkInterface) -> "Interface":
             and ConnectionStateFlags.IP6_READY in inet.connection.state_flags
         )
 
+        if inet.settings and inet.settings.connection:
+            mdns = inet.settings.connection.mdns
+            lldmp = inet.settings.connection.lldmp
+        else:
+            mdns = None
+            lldmp = None
```
The bot is right, this should be corrected as well.

sairon commented on supervisor/dbus/const.py:
```diff
@@ -296,6 +296,14 @@ class MulticastProtocolEnabled(StrEnum):
 
 
+class MulticastDnsValue(IntEnum):
+    """Connection MulticastDNS (mdns/lldmp) values."""
```
```suggestion
    """Connection MulticastDNS (mdns/llmnr) values."""
```

sairon commented on supervisor/host/configuration.py:
```diff
@@ -145,6 +154,13 @@ def from_dbus_interface(inet: NetworkInterface) -> "Interface":
 
+        if inet.settings and inet.settings.connection:
+            mdns = inet.settings.connection.mdns
+            lldmp = inet.settings.connection.lldmp
```
```suggestion
            llmnr = inet.settings.connection.llmnr
```

sairon commented on supervisor/host/configuration.py:
```diff
@@ -145,6 +154,13 @@ def from_dbus_interface(inet: NetworkInterface) -> "Interface":
+            lldmp = inet.settings.connection.lldmp
+        else:
+            mdns = None
+            lldmp = None
```
```suggestion
            llmnr = None
```

sairon commented on supervisor/host/configuration.py:
```diff
@@ -181,6 +197,8 @@ def from_dbus_interface(inet: NetworkInterface) -> "Interface":
             Interface._map_nm_wifi(inet),
             Interface._map_nm_vlan(inet),
+            Interface._map_nm_multicast_dns(mdns),
+            Interface._map_nm_multicast_dns(lldmp),
```
```suggestion
            Interface._map_nm_multicast_dns(llmnr),
```

agners commented:
@t3hk0d3 what is the state of this PR from your point of view? If you feel it is ready to be merged, press the "Ready for review" button so it is no longer in draft state.

agners commented on supervisor/api/network.py:
```diff
@@ -90,6 +93,8 @@
         vol.Optional(ATTR_IPV6): _SCHEMA_IPV6_CONFIG,
         vol.Optional(ATTR_WIFI): _SCHEMA_WIFI_CONFIG,
         vol.Optional(ATTR_ENABLED): vol.Boolean(),
+        vol.Optional(ATTR_MDNS): vol.Coerce(MulticastDnsMode),
+        vol.Optional(ATTR_LLMNR): vol.Coerce(MulticastDnsMode),
```
The option should be named something like `multicast_name_resolution`. But I am a bit tied on that one. I *guess* most cases are fine with it affecting both options, but I like that the options actually say what they mean (mDNS/LLMNR). I almost feel that introducing a new term ("Multicast name resolution", which we then need to explain what that entails, makes things more complicated than just have the two options separately :shrug:

agners replied:
Resolved, the latest version went the suggested approach of using separate options :+1: .

t3hk0d3 commented:
> @t3hk0d3 what is the state of this PR from your point of view? If you feel it is ready to be merged, press the "Ready for review" button so it is no longer in draft state.

I am quite busy lately. I still need to test API changes.

I've posted MR draft so other people would be aware of it, and also could give some early feedback.

t3hk0d3 commented:
beep boop

nikeshpurohit commented:
keep open, bot

chriscrowe commented:
I would absolutely love to see this get merged -- mDNS packet storms have been plaguing my network for months (maybe years). Please merge for the sake of my family's sanity and my marriage 🤣

agners commented on supervisor/dbus/network/setting/generate.py:
```diff
@@ -134,6 +135,16 @@ def _get_ipv6_connection_settings(ipv6setting) -> dict:
+        MulticastDnsMode.ANNOUNCE: MulticastDnsValue.ANNOUNCE,
+    }
+
+    return int(mapping[mode] if mode else MulticastDnsValue.ANNOUNCE)
```
There is actually also the mode -1 (default). Since we created the connections with the hardcoded value 2 (announce) so far, it probably hardly ever happens.

I wonder if we actually should remove the hard coded announce (2) today and use default (-1) (the default is announce (2) too, at least on HAOS since https://github.com/home-assistant/operating-system/pull/986, but it feels cleaner). Anyhow, this should be a separate PR, possible only once we start to officially deprecate old operating system versions.

But ideally we introduce the default option here already, since Supervisor might also encounter such a connection setting today. I guess this should be straight forward, essentially add `Default` with a value of `-1` to `MulticastDnsMode` and `MulticastDnsValue`.

agners reviewed (changes requested):
@t3hk0d3 thanks for the update! Would be cool if we can get this through to the finish line soon. There are some pytest issues as well. Let me know if you need any help.

robloh commented:
Not stale

t3hk0d3 commented:
Finally got some spare time to finish this PR. Sorry for long waiting everyone!

New changes:
- Fixed tests and tested changes in my local setup (built and installed supervisor on my local RPI)
- Changed default mDNS/LLMNR mode for `default` (-1) (as per @agners suggestion)
- Changed default mDNS/LLMNR mode for VLAN interfaces created using API to use `resolve-only` mode by default (can be overriden by setting `mdns`/`llmnr` keys in API request)

agners commented on supervisor/api/network.py:
```diff
@@ -300,6 +311,16 @@ async def create_vlan(self, request: web.Request) -> None:
         vlan_config = VlanConfig(vlan, interface.name)
 
+        # Set MDNS mode for VLANs to RESOLVE by default
+        mdns_mode = MulticastDnsMode.RESOLVE
+        llmnr_mode = MulticastDnsMode.RESOLVE
```
This essentially makes mDNS and LLMNR unusable for those of us who respect the intention of link-local protocols (which is, to operate them link local only, see also https://github.com/home-assistant/supervisor/issues/5435#issuecomment-2500411115).

For those who run reflector it should be easy to switch the switch. But let's go with default configuration by default.

```suggestion
        mdns_mode = MulticastDnsMode.DEFAULT
        llmnr_mode = MulticastDnsMode.DEFAULT
```

agners commented on supervisor/dbus/network/setting/generate.py:
```diff
@@ -189,13 +202,16 @@ def get_connection_from_interface(
             CONF_ATTR_CONNECTION_TYPE: Variant("s", iftype),
-            CONF_ATTR_CONNECTION_LLMNR: Variant("i", 2),
-            CONF_ATTR_CONNECTION_MDNS: Variant("i", 2),
+            CONF_ATTR_CONNECTION_LLMNR: Variant("i", llmnr),
+            CONF_ATTR_CONNECTION_MDNS: Variant("i", mdns),
```
Variant probably can handle `IntEnum`'s, but better to be explicit:

```suggestion
            CONF_ATTR_CONNECTION_LLMNR: Variant("i", int(llmnr)),
            CONF_ATTR_CONNECTION_MDNS: Variant("i", int(mdns)),
```

agners commented on supervisor/dbus/network/setting/generate.py:
```diff
@@ -163,6 +165,17 @@ def _get_ipv6_connection_settings(
     return ipv6
 
 
+def _map_mdns_setting(mode: MulticastDnsMode | None) -> int:
```
```suggestion
def _map_mdns_setting(mode: MulticastDnsMode | None) -> MulticastDnsValue:
```

agners commented on supervisor/host/configuration.py:
```diff
@@ -340,3 +353,14 @@ def _map_nm_vlan(inet: NetworkInterface) -> VlanConfig | None:
+    @staticmethod
+    def _map_nm_multicast_dns(mode: int | None) -> MulticastDnsMode | None:
+        mapping = {
+            MulticastDnsValue.DEFAULT: MulticastDnsMode.DEFAULT,
+            MulticastDnsValue.OFF: MulticastDnsMode.OFF,
+            MulticastDnsValue.RESOLVE: MulticastDnsMode.RESOLVE,
+            MulticastDnsValue.ANNOUNCE: MulticastDnsMode.ANNOUNCE,
```
```suggestion
            MulticastDnsValue.DEFAULT.value: MulticastDnsMode.DEFAULT,
            MulticastDnsValue.OFF.value: MulticastDnsMode.OFF,
            MulticastDnsValue.RESOLVE.value: MulticastDnsMode.RESOLVE,
            MulticastDnsValue.ANNOUNCE.value: MulticastDnsMode.ANNOUNCE,
```

t3hk0d3 replied:
tbh i would also extract this mapping into constant to avoid runtime allocation overhead, although i am not a Python-expert at all, so it can be just a worthless micro-optimization

agners reviewed (changes requested):
Thanks for the updated PR! My main gripe is the default. Other than that, looks good to me!

t3hk0d3 commented:
@agners 
> Thanks for the updated PR! My main gripe is the default. Other than that, looks good to me!

Thank you for checking! I've addressed your comments.

agners commented on supervisor/host/configuration.py:
```diff
@@ -340,3 +360,7 @@ def _map_nm_vlan(inet: NetworkInterface) -> VlanConfig | None:
+
+    @staticmethod
+    def _map_nm_multicast_dns(mode: int | None) -> MulticastDnsMode | None:
+        return _MULTICAST_DNS_VALUE_MODE_MAPPING[mode] if mode else None
```
0 evaluates to false here, but is a valid mode with a mapping. I think we can simply use `get` here:

```suggestion
        return _MULTICAST_DNS_VALUE_MODE_MAPPING.get(mode)
```

t3hk0d3 replied:
Good catch. Fixed

agners commented on supervisor/dbus/network/setting/__init__.py:
```diff
@@ -234,6 +234,8 @@ async def reload(self):
                 interface_name=data[CONF_ATTR_CONNECTION].get(
                     CONF_ATTR_CONNECTION_INTERFACE_NAME
                 ),
+                mdns=data[CONF_ATTR_CONNECTION].get(CONF_ATTR_CONNECTION_MDNS),
+                llmnr=data[CONF_ATTR_CONNECTION].get(CONF_ATTR_CONNECTION_LLMNR),
```
By testing, it seems when using the default (-1) the attribute is not even saved. So this get will return None, which then causes the API to show `llmnr: null` and `mdns: null`.

I haven't tried what happens when I try to update the network config, we probably have some safeguard to not actually write null back. But anyhow, we should just assume default if the property is missing, so:

```suggestion
                mdns=data[CONF_ATTR_CONNECTION].get(CONF_ATTR_CONNECTION_MDNS, -1),
                llmnr=data[CONF_ATTR_CONNECTION].get(CONF_ATTR_CONNECTION_LLMNR, -1),
```

We don't really use the NM settings specific enums in this code, but I think we maybe should use `MulticastDnsValue.DEFAULT.value` here.

t3hk0d3 replied:
Good catch. Added `MulticastDnsValue.DEFAULT.value` here.

t3hk0d3 commented:
@agners
Thanks for checking and approving this PR!
Somehow I can't merge it - GitHub is either waiting for @sairon 's approval or have some maintainer to merge it.
